tasks/get_sql_logins.ps1

```diff
@@ -0,0 +1,254 @@
+      exitcode     = 1
+    }
+  }
+}
+
+function Select-LoginName {
+  param(
+    [PSObject]$login,
+    [string[]]$namesToMatch,
+    [switch]$exact_match
+  )
+
+  <#
+    This function takes a single SQLServer login object and compares it against
+    the list of names passed into the -login_name parameter of the script to
+    determine if this is a login the user is interested in seeing. If it does
+    not pass the filter represented by that parameter the login is discarded.
+  #>
+
+  foreach ($paramLogin in [string[]]$namesToMatch) {
+    if ($exact_match) {
+      if ($paramLogin -eq $login.name) {
+        Write-Output $login
+      }
+    }
+    else {
+      # Match is a regex operator, and it doesn't like the '\' in domain names.
+      if ($login.name.replace('\', '') -match $paramLogin.replace('\', '')) {
+        Write-Output $login
+      }
+    }
+  }
+}
+
+function Get-SQLInstances {
+  param(
+    [string[]]$instance_name
+  )
+
+  $instancesHolder = @()
+  $stringsToReturn = @()
+
+  # The default instance is referred to in its service name as MSSQLSERVER. This
+  # leads many SQLSERVER people to refer to it as such. They will also connect
+  # to it using just a '.'. None of these are it's real name. Its real instance
+  # name is just the machine name. A named instances real name is the machine
+  # name a '\' and the instance name. This little foreach ensures that we are
+  # referring to these instances by their real names so that proper filtering
+  # can be done.
+
+  foreach ($name in $instance_name) {
+    switch ($name) {
+      {($_ -eq 'MSSQLSERVER') -or ($_ -eq '.')} { $instancesHolder += $env:COMPUTERNAME }
+      {$_ -eq $env:COMPUTERNAME} { $instancesHolder += $_ }
+      {$_ -notmatch '\\'} { $instancesHolder += "$env:COMPUTERNAME\$_" }
+      default {$instancesHolder += $name}
+    }
+  }
+
+  $instanceStrings = (Get-ItemProperty 'HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server').InstalledInstances
+
+  # The registry key does not return the real instance names. Again we must
+  # normalize these names into their real names so that comparisons can be done
+  # properly.
+
+  foreach ($string in $instanceStrings) {
+    switch ($string) {
+      'MSSQLSERVER' { $string = $env:COMPUTERNAME }
+      Default {$string = "$env:COMPUTERNAME\$string"}
+    }
+
+    if ((-not [string]::IsNullOrEmpty($instancesHolder))-and(-not [string]::IsNullOrWhiteSpace($instancesHolder))) {
+      foreach ($instance in $instancesHolder) {
+        if ($instance -eq $string) {
+          $stringsToReturn += $string
+        }
+      }
+    }
+    else {
+      $stringsToReturn += $string
+    }
+  }
+
+  if($stringsToReturn.count -gt 0){
+    Write-Output $stringsToReturn
+  } else {
+    throw "No instances were found by the name(s) $instance_name"
+  }
+}
+
+function Get-ServerObject {
+  param(
+    [string]$instance
+  )
+
+  [System.Reflection.Assembly]::LoadWithPartialName("Microsoft.SqlServer.Smo") | Out-Null
```

Comment:
`Out-Null` can be pretty slow, might be worth using `> $null` or `[void]` - but this is also only one action performed once so maybe not.

Reply:
Done.